src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -274,40 +293,100 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
+            var newTrees = new NodeStateTable<SyntaxTree>.Builder();
+            foreach ((var tree, var treeState) in previousTrees)
+            {
+                newTrees.AddEntriesFromPreviousTable(previousTrees, syntaxTrees.Contains(tree) ? EntryState.Cached : EntryState.Removed);
+                syntaxTrees.Remove(tree);
+            }
+            foreach (var tree in syntaxTrees)
+            {
+                newTrees.AddEntries(ImmutableArray.Create(tree), EntryState.Added);
+            }
+            var syntaxTreeTable = newTrees.ToImmutableAndFree();
+            driverStateBuilder.SetInputState(SharedInputNodes.SyntaxTrees, syntaxTreeTable);
+            syntaxTrees.Free();
+
+            // PROTOTYPE(source-generators): Can we do this inline as we update the trees above?
+            if (syntaxInputNodes.Count > 0 || receiverCount > 0)
+            {
+                var builders = syntaxInputNodes.SelectAsArray(n => n.GetBuilder(_state.StateTable));
+                foreach ((var tree, var treeState) in driverStateBuilder.GetLatestStateTableForNode(SharedInputNodes.SyntaxTrees))
                 {
-                    var root = tree.GetRoot(cancellationToken);
+                    if (treeState == EntryState.Removed)
+                    {
+                        // building a new state, so don't need to worry about the removed.
+                        continue;
+                    }
+
+                    // PROTOTYPE(source-generators): It would be nice to somehow collect all the results up front, then pass them to the builder
+                    //                               so we don't have to worry about the generic and have the extra interface.
+
                     var semanticModel = compilation.GetSemanticModel(tree);
+                    Lazy<SyntaxNode> root = new Lazy<SyntaxNode>(() => tree.GetRoot(cancellationToken));
 
-                    // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
-                    for (int i = 0; i < walkerBuilder.Count; i++)
+                    if (treeState == EntryState.Cached)
                     {
-                        var walker = walkerBuilder[i];
-                        if (walker is object)
+                        foreach (var b in builders)
                         {
-                            try
-                            {
-                                walker.VisitWithModel(semanticModel, root);
-                            }
-                            catch (Exception e)
+                            b.AddFilterFromPreviousTable(semanticModel);
+                        }
+                    }
+                    else
+                    {
+                        // run a walk for each filter, and save the results
+                        var results = IncrementalGeneratorSyntaxWalker.Run(root.Value, builders);
```

Comment:
>var results [](http://example.com/codeflow?start=24&length=11)

Consider passing in a pooled `ArrayBuilder<ImmutableArray<SyntaxNode>>` to `Run()` rather than having `Run()` allocate an `ImmutableArray<ImmutableArray<SyntaxNode>>`.